extensions-contrib/azure-extensions/src/main/java/org/apache/druid/data/input/azure/AzureInputSource.java

```diff
@@ -0,0 +1,117 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.data.input.azure;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import org.apache.druid.data.input.InputSplit;
+import org.apache.druid.data.input.impl.CloudObjectInputSource;
+import org.apache.druid.data.input.impl.CloudObjectLocation;
+import org.apache.druid.data.input.impl.SplittableInputSource;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.storage.azure.AzureCloudBlobDruidToCloudObjectLocationConverter;
+import org.apache.druid.storage.azure.AzureCloudBlobIterableFactory;
+import org.apache.druid.storage.azure.AzureStorage;
+import org.apache.druid.storage.azure.CloudBlobDruid;
+
+import javax.annotation.Nullable;
+import java.net.URI;
+import java.util.List;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class AzureInputSource extends CloudObjectInputSource<AzureEntity>
+{
+  static final int MAX_LISTING_LENGTH = 1024;
+  static final String SCHEME = "azure";
+
+  private final Logger log = new Logger(AzureInputSource.class);
+  private final AzureStorage storage;
+  private final AzureEntityFactory entityFactory;
+  private final AzureCloudBlobIterableFactory azureCloudBlobIterableFactory;
+  private final AzureCloudBlobDruidToCloudObjectLocationConverter azureCloudBlobToLocationConverter;
+
+  @JsonCreator
+  public AzureInputSource(
+      @JacksonInject AzureStorage storage,
+      @JacksonInject AzureEntityFactory entityFactory,
+      @JacksonInject AzureCloudBlobIterableFactory azureCloudBlobIterableFactory,
+      @JacksonInject AzureCloudBlobDruidToCloudObjectLocationConverter azureCloudBlobToLocationConverter,
+      @JsonProperty("uris") @Nullable List<URI> uris,
+      @JsonProperty("prefixes") @Nullable List<URI> prefixes,
+      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects
+  )
+  {
+    super(SCHEME, uris, prefixes, objects);
+    this.storage = Preconditions.checkNotNull(storage, "AzureStorage");
+    this.entityFactory = Preconditions.checkNotNull(entityFactory, "AzureEntityFactory");
+    this.azureCloudBlobIterableFactory = Preconditions.checkNotNull(
+        azureCloudBlobIterableFactory,
+        "AzureCloudBlobIterableFactory"
+    );
+    this.azureCloudBlobToLocationConverter = Preconditions.checkNotNull(azureCloudBlobToLocationConverter, "AzureCloudBlobToLocationConverter");
+  }
+
+  @Override
+  protected AzureEntity createEntity(InputSplit<CloudObjectLocation> split)
+  {
+    return entityFactory.create(split.get());
```

Comment:
Out of curiosity, what's the benefit of using the factory pattern here?

Reply:
It makes it a lot easier to test